src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs

```diff
@@ -0,0 +1,317 @@
+
+        [Guid("9B164E40-C3A2-4363-9BC5-EB4039DEF653")]
+        private class SVsSettingsPersistenceManager { }
+
+        enum CompatibilityLevel
+        {
+            Supported = 0,
+            Partial = 1,
+            NotSupported =2
+        }
+        [ImportingConstructor]
+        public DotNetCoreProjectCompatibilityDetector([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, Lazy<IProjectServiceAccessor> projectAccessor, 
+                                                      Lazy<IDialogServices> dialogServices, Lazy<IProjectThreadingService> threadHandling)
+        {
+            _serviceProvider = serviceProvider;
+            ProjectServiceAccessor = projectAccessor;
+            DialogServices = dialogServices;
+            ThreadHandling = threadHandling;
+        }
+
+        public async Task InitializeAsync()
+        {
+            await ThreadHandling.Value.SwitchToUIThread();
+
+            // Do nothing, don't hook into events if not the release channel
+            if(IsReleaseChannel())
+            {
+                var solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                Verify.HResult(solution.AdviseSolutionEvents(this, out _solutionCookie));
+            }
+        }
+
+        private Lazy<IProjectServiceAccessor> ProjectServiceAccessor { get; set; }
+        private Lazy<IDialogServices> DialogServices { get; set; }
+        private Lazy<IProjectThreadingService> ThreadHandling { get; set; }
+
+        private readonly IServiceProvider _serviceProvider;
+        private uint _solutionCookie = VSConstants.VSCOOKIE_NIL;
+        private bool _solutionOpened;
+        private CompatibilityLevel _compatibilityLevelWarnedForThisSolution = CompatibilityLevel.Supported;
+
+        public void Dispose()
+        {
+            ThreadHandling.Value.VerifyOnUIThread();
+
+            if(_solutionCookie != VSConstants.VSCOOKIE_NIL)
+            {
+                var solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if(solution != null)
+                {
+                    Verify.HResult(solution.UnadviseSolutionEvents(_solutionCookie));
+                    _solutionCookie = VSConstants.VSCOOKIE_NIL;
+                }
+            }
+        }
+
+        public int OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)
+        {
+            // Only check this project if the solution is opened and we haven't already warned at the maximum level. Note that fAdded
+            // is true for both add and a reload of an unloaded project
+            if(_solutionOpened && fAdded == 1 && _compatibilityLevelWarnedForThisSolution != CompatibilityLevel.NotSupported)
+            {
+                UnconfiguredProject project = pHierarchy.AsUnconfiguredProject();
+                if (project != null)
+                {
+                    ThreadHandling.Value.JoinableTaskFactory.StartOnIdle(async () =>
```

Comment:
Why double-hop to idle and then to the threadpool?

Reply:
That's a hold over from when my early version of the code where it used IVsHiearcny to get the TFM. I've removed it in both places.